feat: slightly better serializer for unit actions and state, added attack action too

Scoppio opened the pull request:
# What it does?

Improves the current TSV Serializer and the actions for the game dataset logger. 
- Adds more fields too for UnitAction and UnitState
- Adds AttackUnitAction 
- and a few new function for EntityFeatureUtils that can gets the percent of armor on each "side" of a unit. Mostly used exclusively for Dataset logger right now, but will be used for other things on the feature extractor/axis calculator for caspar at a later date.
- Cover all data that was missing for custom units, and the info that was missing and I had to extract directly from MegaMek to run my model trainer.

Scoppio commented on megamek/src/megamek/ai/dataset/UnitAction.java:
```diff
@@ -27,80 +27,89 @@
+    public static UnitAction fromMovePath(MovePath movePath) {
         Entity entity = movePath.getEntity();
-        double chanceOfFailure = SharedUtility.getPSRList(movePath).stream().map(psr -> psr.getValue() / 36d).reduce(0.0, (a, b) -> a * b);
+        double chanceOfFailure = SharedUtility.getPSRList(movePath).stream()
```
I am very dumb...

Scoppio commented on megamek/src/megamek/ai/dataset/UnitAction.java:
```diff
@@ -27,80 +27,89 @@
-        double chanceOfFailure = SharedUtility.getPSRList(movePath).stream().map(psr -> psr.getValue() / 36d).reduce(0.0, (a, b) -> a * b);
+        double chanceOfFailure = SharedUtility.getPSRList(movePath).stream()
+                                       .map(psr -> psr.getValue() / 36d)
+                                       .reduce(0.0, (a, b) -> a * b);
```
```suggestion
                                       .reduce(1.0, (a, b) -> a * b);
```

rjhancock commented on megamek/src/megamek/ai/dataset/UnitAttackAction.java:
```diff
@@ -0,0 +1,130 @@
+/*
+ * Copyright (C) 2017-2025 The MegaMek Team. All Rights Reserved.
+ *
+ * This file is part of MegaMek.
```
If a new-ish file, correct the copyright to correspond to when it was first merged.

rjhancock commented on megamek/src/megamek/ai/dataset/UnitAttackField.java:
```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (C) 2017-2025 The MegaMek Team. All Rights Reserved.
+ *
+ * This file is part of MegaMek.
+ *
```
Same here for copyright.

rjhancock commented on megamek/src/megamek/ai/dataset/UnitAttackSerializer.java:
```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright (C) 2017-2025 The MegaMek Team. All Rights Reserved.
+ *
+ * This file is part of MegaMek.
+ *
```
And here....

rjhancock commented on megamek/src/megamek/ai/utility/EntityFeatureUtils.java:
```diff
@@ -0,0 +1,321 @@
+/*
+ * Copyright (C) 2017-2025 The MegaMek Team. All Rights Reserved.
+ *
+ * This file is part of MegaMek.
+ *
+ * MegaMek is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License (GPL),
```
And here...

rjhancock commented on megamek/unittests/megamek/common/EntityFeatureUtilsTest.java:
```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright (C) 2017-2025 The MegaMek Team. All Rights Reserved.
+ *
+ * This file is part of MegaMek.
+ *
+ * MegaMek is free software: you can redistribute it and/or modify
```
And here...

Sleet01 commented on megamek/src/megamek/ai/dataset/UnitAction.java:
```diff
@@ -27,80 +28,89 @@
+    public static UnitAction fromMovePath(MovePath movePath) {
         Entity entity = movePath.getEntity();
-        double chanceOfFailure = SharedUtility.getPSRList(movePath).stream().map(psr -> psr.getValue() / 36d).reduce(0.0, (a, b) -> a * b);
+        double chanceOfFailure = SharedUtility.getPSRList(movePath).stream()
```
Why is this being pulled out here?  It should live in UnitAction if UnitAction cares about it.

Sleet01 replied:
Same for `steps`, actually.

Sleet01 replied:
See above; I think this can be made less boiler-platey (or that the boilerplate can be pushed down a layer).

Scoppio replied:
UnitAction does not keep the movepath, only primitive values that represents it for logging.

Those two variables vould be down there, but are up here, no big reason why.

Sleet01 commented on megamek/src/megamek/ai/dataset/UnitAction.java:
```diff
@@ -27,80 +28,89 @@
-public record UnitAction(int id, int teamId, int playerId, String chassis, String model, int facing, int fromX, int fromY, int toX, int toY, int hexesMoved, int distance, int mpUsed,
-                         int maxMp, double mpP, double heatP, double armorP, double internalP, boolean jumping, boolean prone,
-                         boolean legal, double chanceOfFailure, List<MovePath.MoveStepType> steps, boolean bot) {
+public record UnitAction(int id, int teamId, int playerId, String chassis, String model, int facing,
+      int fromX, int fromY, int toX, int toY, int hexesMoved, int distance,
+      int mpUsed, int maxMp, double mpP, double heatP, double armorP, double internalP,
+      boolean jumping, boolean prone, boolean legal, double chanceOfFailure,
+      List<MovePath.MoveStepType> steps, boolean bot,
+      boolean hasEcm, int armor, int internal, int bv, int maxRange, int totalDamage,
+      float armorFrontP, float armorLeftP, float armorRightP, float armorBackP, UnitRole role,
+      List<Integer> weaponDamageFacingShortMediumLongRange) {
```
While I understand that this is a record, I would highly recommend adding more ergonomic constructors similar to the static method below.
One that accepts just `entity` and `movePath`, similar to the static method, but then populates the values and calls the base constructor, would be useful.

Sleet01 replied:
Actually, looking at the code again, this would just be extra work for no benefit.
But I'd like to advise that 8 lines of constructor parameters is going to be fragile and prone to breakage.

Scoppio replied:
it is a data objet that represents the entire TSV line, there isnt much more to do here other than break into multiple objects, which also would not solve the problem.
Record here is good because it is immutable, and it being prone to breakage, well, I literally can't do better.

Sleet01 reviewed (approved):
After some discussion on Discord I have changed my mind on the records approach.
Everything looks good, excited to see the next steps!

Scoppio commented:
> After some discussion on Discord I have changed my mind on the records approach. Everything looks good, excited to see the next steps!

And I changed my mind... this is really annoying to improve upon, so I am throwing the towel and going for the thing I didnt wanted to do before but is possibly the correct solution, using an EnumHashMap  to hold the values and use reflection to serialize the object.

Sleet01 reviewed (approved):
Looks good!  Nice job compacting some of that code down.

Scoppio commented on megamek/src/megamek/common/GameDatasetLogger.java:
```diff
@@ -239,142 +198,30 @@ private void append(Game game, AbstractAttackAction attackAction, boolean withHe
-            );
+            appendToFile(unitAttackSerializer.serialize(unitAttackAction));
+        } catch (Exception ex) {
+            logger.error(ex, "Error logging Attack unit action");
         }
```
The original code is correct
